### PR TITLE
statistics: delete extended stats cache item in current tidb synchronously

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23065

Problem Summary:

Not a problem, change the extended stats removal to a more intuitive behavior.

### What is changed and how it works?

What's Changed:

- add `minorVersion` for `statsCache`, and check it when updating stats cache;
- remove extended stats cache item in `alter table drop stats_extended`;

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 

- Delete extended stats cache item in current tidb synchronously

### CODE REVIEW

File: statistics/handle/handle.go [hunk 1]
```diff
@@ -56,8 +56,10 @@ const (
 type statsCache struct {
 	tables map[int64]*statistics.Table
 	// version is the latest version of cache.
-	version  uint64
-	memUsage int64
+	version uint64
+	// minorVersion is to differentiate the cache when version is same.
+	minorVersion uint64
```
Comment (qw4990) [inline statistics/handle/handle.go:61]: It seems confused. If I update the stats item, which one should be increased? Are there any principles or rules we should obey?

Comment (winoros) [reply statistics/handle/handle.go:61]: Could we explain how we use it here?

Comment (eurekaka) [reply statistics/handle/handle.go:61]: Notice that for function `func (sc statsCache) update()`, a param `newVersion` is required. There are 3 callers of this function in total except extended stats component:
1) `LoadNeededHistograms` would insert column histograms into stats cache;
2) `GetPartitionStats` would insert a pseudo table stats;
3) the `Update()` procedure of stats;

After 1) and 2), the stats cache is actually at the same version as the one before the modification, so `newVersion` for `func (sc statsCache) update()` in these 2 cases are same as the old one, i.e, there is no version bump. Only the third case would bump the stats version if new stats are read from storage.

There would be lost of modifications if 1) and 2) happens concurrently(extended stats modification also belongs to this group then), that's why I introduce `minorVersion`, to differentiate stats caches which have trivial differences but have same version.

Comment (eurekaka) [reply statistics/handle/handle.go:61]: > Are there any principles or rules we should obey?

We don't need to care about this `minorVersion` when writing code, if we use same `version` in `update()`, `minorVersion` would automatically be increased internally.

Comment (qw4990) [reply statistics/handle/handle.go:61]: After discussing it with @eurekaka offline, I finally got it... Please add more comments about this variable, especially a case.

Comment (time-and-fate) [reply statistics/handle/handle.go:61]: `(h *Handle) UpdateErrorRate` and `(h *Handle) UpdateStatsByLocalFeedback` also used `updateStatsCache()` without version bump, seems they also need modified?

**Comment by eurekaka** [comment]
@qw4990 @winoros I have added more comments for the `minorVersion`, PTAL.

**Comment by time-and-fate** [comment]
`(h *Handle) UpdateErrorRate` and `(h *Handle) UpdateStatsByLocalFeedback` also used `updateStatsCache()` without version bump, seems they also need modified?

File: statistics/handle/handle.go [hunk 2]
```diff
@@ -55,9 +55,18 @@ const (
...
-	memUsage int64
+	// version is the latest version of cache. It is bumped when new records of `mysql.stats_meta` are loaded into cache.
+	version uint64
+	// minorVersion is to differentiate the cache when the version is unchanged while the cache contents are
+	// modified indeed. This can happen when we load extra column histograms into cache, or when we modify the cache with
+	// statistics feedbacks, etc. We cannot bump the version then because no new changes of `mysql.stats_meta` are loaded,
+	// while the override of statsCache is in a copy-on-write way, to make sure the statsCache is unchanged during the
+	// the interval of 'copy' and 'write' by others, every 'write' should bump / check this minorVersion if the version keeps
+	// unchanged.
+	// This bump / check logic is encaptured in `statsCache.update` and `updateStatsCache`, callers don't need to care
```
Comment (time-and-fate) [inline statistics/handle/handle.go:66]: encaptured => encapsulated?

File: statistics/handle/handle.go [hunk 3]
```diff
@@ -516,20 +525,24 @@ func (h *Handle) GetPartitionStats(tblInfo *model.TableInfo, pid int64) *statist
 	return tbl
 }
 
-func (h *Handle) updateStatsCache(newCache statsCache) {
+func (h *Handle) updateStatsCache(newCache statsCache) (updated bool) {
```
Comment (time-and-fate) [inline statistics/handle/handle.go:528]: I think we should remind other people in the comments that this function may fail when `version` is not changed and retry may be needed.

**Review by time-and-fate — APPROVED** [review]
/lgtm

**Review by qw4990 — APPROVED** [review]

**Pull request merged**